pypeit/coadd2d.py

```diff
@@ -1421,31 +1484,52 @@ def get_brightest_obj(self, specobjs_list, nslits):
 
         objid = np.zeros(nexp, dtype=int)
         snr_bar = np.zeros(nexp)
         # norders = specobjs_list[0].ech_orderindx.max() + 1
         for iexp, sobjs in enumerate(specobjs_list):
+            msgs.info("Working on exposure {}".format(iexp))
             uni_objid = np.unique(sobjs.ECH_OBJID)
             nobjs = len(uni_objid)
-            order_snr = np.zeros((nslits, nobjs))
+            # order_snr = np.zeros((nslits, nobjs))
+            order_snr = np.full((nslits, nobjs), np.nan)
             for iord in range(nslits):
                 for iobj in range(nobjs):
                     ind = (sobjs.ECH_ORDERINDX == iord) & (sobjs.ECH_OBJID == uni_objid[iobj])
-                    # ignore objects for which optimal extraction could not be performed (i.e., OPT_COUNTS is None)
-                    if np.any(sobjs[ind][0].OPT_COUNTS):
+                    # check if OPT_COUNTS is available
+                    if sobjs[ind][0].OPT_COUNTS is not None and sobjs[ind][0].OPT_COUNTS_IVAR is not None \
+                            and sobjs[ind][0].OPT_WAVE is not None and sobjs[ind][0].OPT_MASK is not None:
                         flux = sobjs[ind][0].OPT_COUNTS
                         ivar = sobjs[ind][0].OPT_COUNTS_IVAR
                         wave = sobjs[ind][0].OPT_WAVE
                         mask = sobjs[ind][0].OPT_MASK
+                        # check if BOX_COUNTS is available
+                    elif sobjs[ind][0].BOX_COUNTS is not None and sobjs[ind][0].BOX_COUNTS_IVAR is not None \
+                             and sobjs[ind][0].BOX_WAVE is not None and sobjs[ind][0].BOX_MASK is not None:
+                        msgs.warn(f'Optimal extraction not available for object {sobjs[ind][0].ECH_OBJID} '
+                                  f'in order {sobjs[ind][0].ECH_ORDER}. Using box extraction.')
+                        flux = sobjs[ind][0].BOX_COUNTS
+                        ivar = sobjs[ind][0].BOX_COUNTS_IVAR
+                        wave = sobjs[ind][0].BOX_WAVE
+                        mask = sobjs[ind][0].BOX_MASK
+                    else:
+                        flux = None
+                        msgs.warn(f'Optimal and Boxcar extraction not available for '
+                                  f'object {sobjs[ind][0].ECH_OBJID} in order {sobjs[ind][0].ECH_ORDER}.')
+                        continue
+                    if flux is not None:
                         rms_sn, weights = coadd.sn_weights(wave, flux, ivar, mask, self.sn_smooth_npix, const_weights=True)
                         order_snr[iord, iobj] = rms_sn
 
             # Compute the average SNR and find the brightest object
-            snr_bar_vec = np.mean(order_snr, axis=0)
-            objid[iexp] = uni_objid[snr_bar_vec.argmax()]
-            snr_bar[iexp] = snr_bar_vec[snr_bar_vec.argmax()]
+            # we use nanmean to ignore the nan value that could still be present in order_snr
+            if not np.all(np.isnan(order_snr)):
```

Comment:
Same comment about nans. I guess another thing to think about is what the desired behavior is if one order is masked. In other words, do we mask that entire exposure, or would we only mask that order? I suspect pypeit breaks anyway if a single order is masked but I'm not sure.

Reply:
I'll check that.